### PR TITLE
Auto-normalize dot_product vectors at index & query

### LINKED ISSUE #98935: Auto-normalize vectors when `dot_product` mapping is used
### Description

We have been clear in the past that `dot_product` requires normalized vectors. Its effectively an optimized cosine.

We should normalize vectors for users when they are indexed into Lucene.

Additionally, recent changes have been merged that require a default "similarity" to be used, we should default to dot_product.

NOTE: This is only for `float` vectors.

### PR DESCRIPTION
**Pull request by benwtrent**
`dot_product` requires vectors to be unit-length. Previously, we would check that vectors were unit-length and throw if they were not. 

Instead, we will now auto-normalize vectors as they are indexed.

`cosine` will continue to behave as usual, not normalizing the vectors.

closes: https://github.com/elastic/elasticsearch/issues/98935

### CODE REVIEW

**Comment by benwtrent** [comment]
PR: https://github.com/elastic/elasticsearch/pull/98946 fixes many of the failing tests. There are multiple tests assuming `index: false` and now they have to specifically set it.

One of the difficulties is v7RestCompatibility. 

If we:
 - index: true
 - dot_product == auto-normalize

The v7RestCompatability tests will fail. Because those tests do not specify `index: false`, mainly because that option didn't exist in v7.

For non-indexed vectors the behavior is exactly the same. But the combination of indexing by default and choosing dot_product breaks those tests.

Do we want to keep the default as `cosine` without auto-normalization? 

Do we think this is truly a breaking restcompat change?

File: docs/reference/mapping/types/dense-vector.asciidoc [hunk 1]
```diff
@@ -159,9 +159,9 @@ Computes the dot product of two vectors. This option provides an optimized way
 to perform cosine similarity. The constraints and computed score are defined
 by `element_type`.
 +
-When `element_type` is `float`, all vectors must be unit length, including both
-document and query vectors. The document `_score` is computed as
-`(1 + dot_product(query, vector)) / 2`.
+When `element_type` is `float`, all vectors are automatically converted be unit length, including both
```
Comment (mayya-sharipova) [inline docs/reference/mapping/types/dense-vector.asciidoc:162]: converted to?

File: server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java [hunk 2]
```diff
@@ -678,6 +679,17 @@ float score(float similarity, ElementType elementType, int dim) {
                     case FLOAT -> (1 + similarity) / 2f;
                 };
             }
+
+            @Override
+            void floatPreprocessing(float[] vector, float squareSum) {
+                if (squareSum == 0) {
+                    throw new IllegalArgumentException("Cannot normalize a zero-length vector");
+                }
```
Comment (mayya-sharipova) [inline server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java:687]: if already (squareSum == 1) should we skip the following lines?

File: server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java [hunk 3]
```diff
@@ -881,12 +892,18 @@ public Query createKnnQuery(float[] queryVector, int numCands, Query filter, Flo
             }
             elementType.checkVectorBounds(queryVector);
 
+            float magnitude = Float.NaN;
             if (similarity == VectorSimilarity.DOT_PRODUCT || similarity == VectorSimilarity.COSINE) {
-                float squaredMagnitude = 0.0f;
-                for (float e : queryVector) {
-                    squaredMagnitude += e * e;
-                }
+                float squaredMagnitude = VectorUtil.dotProduct(queryVector, queryVector);
```
Comment (mayya-sharipova) [inline server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java:897]: Nice optimization!

File: server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java [hunk 4]
```diff
@@ -881,12 +892,18 @@ public Query createKnnQuery(float[] queryVector, int numCands, Query filter, Flo
             }
             elementType.checkVectorBounds(queryVector);
 
+            float magnitude = Float.NaN;
```
Comment (mayya-sharipova) [inline server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java:895]: Seems that we can get rid of this line

**Comment by benwtrent** [comment]
For now, I will remove the default to `dot_product` part of this change. We will have to discuss that further.

**Review by mayya-sharipova — APPROVED** [review]
@benwtrent Great simplifications for users! Thanks Ben!

**Comment by benwtrent** [comment]
@elasticmachine update branch

**Comment by benwtrent** [comment]
@elasticmachine update branch

**Pull request merged**